fw/rbcx-coprocessor/src/CdcUartTunnel.cpp

```diff
@@ -3,19 +3,113 @@
 
 #include "stm32f1xx_hal.h"
 #include "stm32f1xx_hal_dma.h"
+#include "stm32f1xx_ll_rcc.h"
 #include "stm32f1xx_ll_usart.h"
 
+#include "FreeRTOS.h"
+#include "stream_buffer.h"
+
 #include "Bsp.hpp"
 #include "CdcUartTunnel.hpp"
 #include "UsbCdcLink.h"
 #include "utils/ByteFifo.hpp"
+#include "utils/Debug.hpp"
+#include "utils/TaskWrapper.hpp"
+
+static DMA_HandleTypeDef dmaRxHandle = {};
+static DMA_HandleTypeDef dmaTxHandle = {};
+
+static volatile bool dmaTxComplete = true;
 
-static DMA_HandleTypeDef dmaRxHandle;
-static DMA_HandleTypeDef dmaTxHandle;
-static ByteFifo<512> rxFifo;
 static std::array<uint8_t, CDC_DATA_SZ> txBuf;
+static ByteFifo<256> rxBuf;
+
+static StaticStreamBuffer_t _rxStreamBufStruct;
+static std::array<uint8_t, 1024> _rxStreamBufStorage;
+static StreamBufferHandle_t rxStreamBuf;
+
+static class : public TaskWrapper<512> {
+public:
+    void start() {
+        TaskWrapper::start(
+            "tunnelRx", configMAX_PRIORITIES - 1, [this]() { rxReader(); });
+    }
+
+private:
+    std::array<uint8_t, CDC_DATA_SZ> m_buf;
+
+    void rxReader() {
+        while (true) {
+            const auto recv = xStreamBufferReceive(
+                rxStreamBuf, m_buf.data(), m_buf.size(), pdMS_TO_TICKS(1));
+            if (recv == 0) {
+                HAL_NVIC_SetPendingIRQ(tunnelUartRxDmaIRQn);
+                continue;
+            }
+            usbWrite(recv);
+        }
+    }
+
+    void usbWrite(size_t size) {
+        // Must be high enough for PC to have enough time to read from the USB,
+        // but low enough so the StreamBuffer won't overflow when PC is not
+        // connected.
+        int attempts = 32;
+
+        size_t sent = 0;
+        while (true) {
+            const auto frame = usbd_ep_write(
+                &udev, CDC_TXD_EP, (void*)(m_buf.data() + sent), size - sent);
+            if (frame > 0) {
+                sent += frame;
+                if (sent >= size) {
+                    return;
+                }
+            }
+
+            if (--attempts == 0)
+                return;
+
+            cdcLinkPoll();
+        }
+    }
```

Comment:
To místo, kde ztrácíme data, je přesunuto ze zápisu do RX bufferu sem - PC si musí USB vyčíst data do 32 poll eventů, jinak je zahodíme.

Znamená to, že když ESP32 něco pošle před tím, než je na PC otevřený port, tak to zahodíme, i když máme pořád místo. Je to protože implementace zahazování tady byla jednoznačně nejjednodušší (FreeRTOS neumí přepisovat stará data ve StreamBufferu) a ESP32 dev modul to dělá taky (zkuste si připojit dev modul k PC, restartovat ho, aby poslal po TX boot message, a pak teprv otevřít v Lorris - nic vám nepřijde).

Reply:
Já pořád nechápu, proč je tento fudging potřeba; *usb_ep_write* nepřepíše EP buffer, dokud ho PC nevyčte. Tady to tedy vytváří backpressure. 

Navíc si nemyslím, že je potřeba mít ty StreamBuffery - jeden buffer je přímo DMA rxFifo, kde můžu tail posouvat, až když ep_write projde.

Stav nepřipojenosti by mělo jít řešit (de)aktivací UARTu podle události detekce (při)odpojení.